### PR TITLE
server: properly check existence of tenant record

### PR DESCRIPTION
**Pull request by knz**
PR extracted from #105441.

Informs #83650.
Part of #98431.
Epic: CRDB-26691

Prior to this patch, the API on the tenant info watcher would return
an "undefined" metadata payload if called prior to the tenant record
being created.

This was most visible in the following scenario:

1. new cluster starts. watcher+rangefeed start successfully (tenant
   table empty)
2. tenant client connects. At this time there is no metadata for
   its tenant ID, so the metadata payload is available but empty.
3. CREATE TENANT is executed for the new tenant.
4. only then (later) the rangefeed introduces the metadata
   into the cache.

This is insufficient for use by the KV tenant client RPCs: there we
only want to accept incoming requests from tenant clients after we
actually have seen metadata for them.

This patch improves the situation by checking whether the tenant
record is present before returning bogus data to the SQL tenant
client.

Simultaneously, we add error handling logic in the SQL tenant client
to handle this case gracefully.

In a mixed-version deployment (with and without this patch applied),
the following behaviors will occur if one starts the SQL tenant server
without a tenant record defined:

- Unpatched server: Late startup error in client with "database 1 does not exist".
- Patched server: Client loops forever, waiting for tenant record.

Behavior when the tenant record is created shortly *after* the SQL
tenant server starts up:

- Unpatched server: Inconsistent / non-deterministic behavior.
- Patched server: Clean startup of client after tenant record appears.

### CODE REVIEW

**Comment by knz** [comment]
Internal discussion: https://cockroachlabs.slack.com/archives/C02HWA24541/p1688559273013399

**Review by stevendanna — COMMENTED** [review]
Thanks for pushing this forward. I gave it an initial read through and left some minor comments.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nkodali, @renatolabs, @srosenberg, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzrO_iEUURbtNAHmIt:-NZzrO_iEUURbtNAHmIu:b-lte32y) ([raw file](https://github.com/cockroachdb/cockroach/blob/00213f1b63d1f0f1a2f6581c5415aa61d1e05f11/pkg/kv/kvclient/kvtenant/setting_overrides.go#L75)):*
@yuzefovich asks

> I don't quite understand this block. `runTenantSettingsSubscription` is called in a single spot in which we derive a child context that is canceled on quiesce, so why do we need first and second `case`s? Also why don't we just exit when the cancellation / quiesce happens? Why don't we exit when non-`io.EOF` error (other than `MissingRecordError`) is `Recv`ed?
___
*[`pkg/kv/kvpb/api.proto` line 3138 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzrGLg5QbljmNcEaLj:-NZzrGLg5QbljmNcEaLk:bc25c60) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a0b6bb432318aca0c3db14a5987fb4aaaa094a2/pkg/kv/kvpb/api.proto#L3138)):*
@yuzefovich asks:
> The second sentence in this paragraph to me implies that creators of `TenantSettingsEvent` must always do all 3 points to keep backwards-compatibility. However, IIUC this is only required until the cluster has been upgraded to 23.2, at which point `TenantSettingsEvent` can be created in any manner, contradicting any of the 3 points. Is that correct?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzraPVDJ948mk90h_d:-NZzraPVDJ948mk90h_e:b-bh9v8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9e10ae32cd1a6b71c73ed548855cbc38794617/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L130)):*
@yuzefovich asks:
> nit: it seems like `nil` settings is only because of `TestingDecoderFn` which is only called in one spot, so I'd refactor the testing fn to pass valid settings object too, in order to remove this non-nil check.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nkodali, @renatolabs, @srosenberg, @stevendanna, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzrO_iEUURbtNAHmIt:-NZzzHUxCSiyMuVOEJfU:b308zr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/00213f1b63d1f0f1a2f6581c5415aa61d1e05f11/pkg/kv/kvclient/kvtenant/setting_overrides.go#L75)):*
> `runTenantSettingsSubscription` is called in a single spot in which we derive a child context that is canceled on quiesce

I hadn't seen that. Simplified.

> why don't we just exit when the cancellation / quiesce happens?

Oversight. Fixed.

> Why don't we exit when non-`io.EOF` error (other than `MissingRecordError`) is `Recv`ed?

This is backward compatibility with current uses. We have roachprod and ORM test suites that require the retry behavior... CI complains loudly when we do the simple exit here, which is why I kept the previous behavior.
___
*[`pkg/kv/kvpb/api.proto` line 3138 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzrGLg5QbljmNcEaLj:-NZzxOGB8PdGWAUFEP4N:bidvbdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a0b6bb432318aca0c3db14a5987fb4aaaa094a2/pkg/kv/kvpb/api.proto#L3138)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

@yuzefovich asks:
> The second sentence in this paragraph to me implies that creators of `TenantSettingsEvent` must always do all 3 points to keep backwards-compatibility. However, IIUC this is only required until the cluster has been upgraded to 23.2, at which point `TenantSettingsEvent` can be created in any manner, contradicting any of the 3 points. Is that correct?
</blockquote></details>

Indeed. Clarified in comment.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzraPVDJ948mk90h_d:-NZzx-twAhhLxxPRfxNt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9e10ae32cd1a6b71c73ed548855cbc38794617/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L130)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

@yuzefovich asks:
> nit: it seems like `nil` settings is only because of `TestingDecoderFn` which is only called in one spot, so I'd refactor the testing fn to pass valid settings object too, in order to remove this non-nil check.
</blockquote></details>

Done.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106414#gh-1257863851:-NZzsxcXA_1wmvZ0Cgaq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9e10ae32cd1a6b71c73ed548855cbc38794617/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L162)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

ReportOrPanic appears to already log the given message at the Error level, given that, is there much value in also logging it as a warning here?
</blockquote></details>

Done.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 237 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/106414#gh-1257890191:-NZztJ9-8Vu004fQ5HpO:b-2fod6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a0b6bb432318aca0c3db14a5987fb4aaaa094a2/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L237)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

The rest of this function doesn't look too onerous, did we see problems without this fast path?
</blockquote></details>

I copied-and-pasted this from server/tenantsettingswatcher. simplified.
___
*[`pkg/server/node.go` line 2059 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106414#gh-1257904409:-NZztNSF23xo8sMDmyQo:b-mlyogk) ([raw file](https://github.com/cockroachdb/cockroach/blob/00213f1b63d1f0f1a2f6581c5415aa61d1e05f11/pkg/server/node.go#L2059)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Perhaps we could use longer variable names here so it is easier to remember which watcher is for what.
</blockquote></details>

Let's do it.
___
*[`pkg/server/tenantsettingswatcher/watcher.go` line 87 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/106414#gh-1257900068:-NZzuGGj-b7iSfWr_XEm:bjlxwc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a0b6bb432318aca0c3db14a5987fb4aaaa094a2/pkg/server/tenantsettingswatcher/watcher.go#L87)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] No need for the func wrapper here I believe.
</blockquote></details>

again, this was a copy paste. simplified.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher_test.go` line 251 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106414#gh-1257881387:-NZzvBJ3DJ9KaVnNzE6J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2ec6946dda8ffa88bfd5d46dfbb20ffba3e3e3a/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher_test.go#L251)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Does this command need to print the whole info struct? Seems like the end-result is mostly having to write out the capabilities twice in the testfile.
</blockquote></details>

Done.

**Review by yuzefovich — APPROVED** [review]
Nice work! :lgtm: although it'd be probably worth it to wait for Steven's approval too.

> Reviewed 2 of 7 files at r2, 36 of 36 files at r6, 19 of 19 files at r7, 14 of 14 files at r8, 20 of 20 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-:-N_0THAt1ifuHZDS0TyN:b8uc7mi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @nkodali, @renatolabs, @srosenberg, and @stevendanna)

___
*[`pkg/kv/kvclient/kvtenant/connector.go` line 963 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-N_0RnFN41yGpS_ZjApN:-N_0RnFN41yGpS_ZjApO:b-uivr1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd1204bf48c835a28f5a16b7acc5f914c6af35f/pkg/kv/kvclient/kvtenant/connector.go#L963)):*
> ```Go
> 	defer c.mu.Unlock()
> 	if c.mu.client == client {
> 		err := c.mu.client.conn.Close() // nolint:grpcconnclose
> ```

nit: should closing the connection be in a separate commit that we want to backport? Or do you plan to backport the whole PR anyway?
___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-NZzrO_iEUURbtNAHmIt:-N_0SN-aB_GL4k6ZA_iX:bg2j5hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/00213f1b63d1f0f1a2f6581c5415aa61d1e05f11/pkg/kv/kvclient/kvtenant/setting_overrides.go#L75)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

> `runTenantSettingsSubscription` is called in a single spot in which we derive a child context that is canceled on quiesce

I hadn't seen that. Simplified.

> why don't we just exit when the cancellation / quiesce happens?

Oversight. Fixed.

> Why don't we exit when non-`io.EOF` error (other than `MissingRecordError`) is `Recv`ed?

This is backward compatibility with current uses. We have roachprod and ORM test suites that require the retry behavior... CI complains loudly when we do the simple exit here, which is why I kept the previous behavior.
</blockquote></details>

Makes sense, thanks.
___
*[`pkg/server/settingswatcher/settings_watcher.go` line 258 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-N_09KUM5rHhM4iX4Wtv:-N_09KUM5rHhM4iX4Wtw:b6pc0u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/17cfc988ef94dfe04320f630b83fc267bb3017d7/pkg/server/settingswatcher/settings_watcher.go#L258)):*
> ```Go
> 		err = errors.NewAssertionErrorWithWrappedErrf(err, "failed to decode settings row %v", kv.Key)
> 		logcrash.ReportOrPanic(ctx, &s.settings.SV, "%w", err)
> 		log.Warningf(ctx, "%v", err)
> ```

nit: ditto for removing redundant logging.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nkodali, @renatolabs, @srosenberg, @stevendanna, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvtenant/connector.go` line 963 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-N_0RnFN41yGpS_ZjApN:-N_3C9BoD_wh4F4t8x2_:b-fdrvt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd1204bf48c835a28f5a16b7acc5f914c6af35f/pkg/kv/kvclient/kvtenant/connector.go#L963)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should closing the connection be in a separate commit that we want to backport? Or do you plan to backport the whole PR anyway?
</blockquote></details>

Not a bad idea! Moving this change to #106576  for discussion.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nkodali, @renatolabs, @srosenberg, @stevendanna, and @yuzefovich)

___
*[`pkg/server/settingswatcher/settings_watcher.go` line 258 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-N_09KUM5rHhM4iX4Wtv:-N_48g9gCjDEzT5cdM9d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17cfc988ef94dfe04320f630b83fc267bb3017d7/pkg/server/settingswatcher/settings_watcher.go#L258)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto for removing redundant logging.
</blockquote></details>

Done.

File: pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go
Comment (stevendanna) [reply pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go:114]: Adding a bit more here. I meant something like:

```golang
dataState, err := mtinfopb.NewDataState(rawDataState)
if err != nil {
    log.Warning...
}
```

Where `NewDataState` is where the bounds check lives.

But, I don't feel strongly about that.

**Review by knz — COMMENTED** [review]
Steven, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-:-N_4hvI5Bykz8uxrRyy_:bfnix6z)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nkodali, @renatolabs, @srosenberg, @stevendanna, and @yuzefovich)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106414#gh-1257869066:-N_4hsdU2usHxMWMWaSc:b2fyll2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2ec6946dda8ffa88bfd5d46dfbb20ffba3e3e3a/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L115)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Adding a bit more here. I meant something like:

```golang
dataState, err := mtinfopb.NewDataState(rawDataState)
if err != nil {
    log.Warning...
}
```

Where `NewDataState` is where the bounds check lives.

But, I don't feel strongly about that.
</blockquote></details>

Thanks, I added a commit to do that.

**Review by yuzefovich — APPROVED** [review]
> Reviewed 33 of 33 files at r10, 13 of 13 files at r11, 9 of 9 files at r12, 15 of 15 files at r13, 7 of 7 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @nkodali, @renatolabs, @srosenberg, and @stevendanna)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 67 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-N_AT0Bi6SHQbC3hUEvX:-N_AT0Bi6SHQbC3hUEvY:b-bssvv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e57d7fb2dffbe48450932b96cf4d18e47ec7ea/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L67)):*
> ```Go
> 
> 	// The remaining columns are stored in the value; we're just interested in the
> 	// info column.
> ```

nit: "just info column" seems out of date now.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106414)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nkodali, @renatolabs, @srosenberg, @stevendanna, and @yuzefovich)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 67 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/106414#-N_AT0Bi6SHQbC3hUEvX:-Na26Ylc3nxnoLRfQPCq:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9e57d7fb2dffbe48450932b96cf4d18e47ec7ea/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L67)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: "just info column" seems out of date now.
</blockquote></details>

Fixed.

**Comment by knz** [comment]
TFYR!

bors r=yuzefovich,stevendanna

**Comment by healthy-pod** [comment]
After this PR was merged, the [multitenant-upgrade started failing](https://github.com/cockroachdb/cockroach/issues/107564) due to:
```
net/http/server.go:3228 ⋮ [-] 528  ‹http: TLS handshake error from redacted_ip:redacted_port: remote error: tls: bad certificate›
```

**Comment by knz** [comment]
followup: #107824

**Pull request merged**
